### PR TITLE
Introducting new option --invariantLevel NUMBER and dropping --omitInvariants

### PR DESCRIPTION
**Pull request by NTillmann**
Release notes: Introducting new option --invariantLevel NUMBER and dropping --omitInvariants

The new default is invariant level 0, which corresponds to the old --omitInvariants.
Level 1 is roughly what we used to do: Check invariants on (derived) abstract values.
Level 2 implements #1180: Checking that all accesses to built-ins do what Prepack expects them to do, including checks for when properties on built-ins are absent.
Level 3 adds checks for internal consistency, basically an internal debug level.

The serializer tests now run with the highest invariant level by default. The added invariants found a few issues that got addressed, including:
- Prepack exposes a TypedArray prototype, which is not a real thing. It's now marked as `__refuseSerialization`, and no invariants are emitted for such things.
- Global variables / properties on the global object are special. Those are not yet handled at level 2.
- Accesses to Prepack magic functions that generally start with `__` should not produce invariant checks.
- Magic code generation for loops should not take into account objects that are `__refuseSerialization`.
- All invariant statements get a unique id to make it easier to locate them.
- I marked some test cases which depend on counting occurrences in the output as "// omit invariants", as the additional invariants increased some such counts.

As part of testing, I also found it necessary to make the --invariantMode more useful; it now also allows specifying nativeLoggingHook which is the preferred way of logging in React Native.

To reduce the number of checks by a few orders of magnitude in practice, each property is only checked on first access. This is tracked by a global variable `__checkedBindings`.

This pull requests incorporates all aspects of the #1709 (which I abandoned).

### CODE REVIEW

**Comment by hermanventer** [comment]
There should be clear guidance to future developers about which level is appropriate for a given invariant check. I expect that 7 levels will be a challenge for such guidance, so consider reducing the number of levels. Also consider the burden on people who will specify the level number on the command line: how do they decide what is an appropriate level?

File: scripts/test-runner.js [hunk 1]
```diff
@@ -266,7 +266,7 @@ function runTest(name, code, options: PrepackOptions, args) {
   let delayUnsupportedRequires = code.includes("// delay unsupported requires");
   if (code.includes("// inline expressions")) options.inlineExpressions = true;
   if (code.includes("// do not inline expressions")) options.inlineExpressions = false;
-  if (code.includes("// omit invariants") || args.verbose) options.omitInvariants = true;
+  options.invariantLevel = code.includes("// omit invariants") || args.verbose ? 0 : 4;
```
Comment (hermanventer) [inline scripts/test-runner.js]: Why 4?

File: src/intrinsics/ecma262/JSON.js [hunk 2]
```diff
@@ -325,11 +325,7 @@ function InternalJSONClone(realm: Realm, val: Value): Value {
     // TODO #1010: NaN and Infinity must be mapped to null.
     return val;
   }
-  if (
-    (val instanceof NumberValue && !isFinite(val.value)) ||
-    val instanceof UndefinedValue ||
-    val instanceof NullValue
-  ) {
+  if (val instanceof NumberValue && !isFinite(val.value)) {
```
Comment (hermanventer) [inline src/intrinsics/ecma262/JSON.js]: There is no discussion about what issue if being resolved by this and this seems a strange PR in which to make such a change.

Comment (NTillmann) [reply src/intrinsics/ecma262/JSON.js]: The additional invariants brought out the bug with the existing serializer test suite that our cloned JSON values where not right. `undefined` doesn't get mapped to `null`. This fixes it to make the test pass. The same code change was already in #1709.

File: scripts/test-runner.js (hunk 1 above)
Comment (NTillmann) [reply scripts/test-runner.js]: 4 roughly corresponds to the old invariant checking default

**Comment by cblappert** [comment]
Make sure that when you land this, you add the flag to internal cases. I agree with Herman that if there are multiple levels, there should be some document/comment/help note that gives guidance on what the different levels mean and what should go in each level.

I wonder if it would be better to have only a few levels of invariant + options to turn on specific checks instead of these indirect verbosity numbers. (e.g. `--invariantLevel "0,undefined,abstract"` or something)

**Comment by NTillmann** [comment]
Before even considering landing this, I am planning to do lots more internal testing.
Right now, I am thinking of reducing the invariant levels to 3:
0: no checking
1: checking of __abstract stuff
2: checking of all accesses to built-ins

File: test/serializer/abstract/InvariantsForNonAbstractProperties.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+// add at runtime:global.a = {p: 42, q: global, r: function() {}};
+// Count of Prepack model invariant violation:5
+(function () {
+    if (global.__assumeDataProperty) __assumeDataProperty(global, "a",
```
Comment (hermanventer) [inline test/serializer/abstract/InvariantsForNonAbstractProperties.js]: Whoa, where did that formatting come from?

**Pull request closed without merging**